tests/fsharp/Compiler/Language/ComputedCollectionRangeLoweringTests.Int32.fs

```diff
@@ -0,0 +1,1826 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace FSharp.Compiler.UnitTests.ComputedCollectionRangeLoweringTests
```

Comment:
Yes, I did have an overflow check, but not a good enough one :)

My check—

```fsharp
while loopVar <= finish && start <= loopVar do
    …
    loopVar <- loopVar + step
```

—worked for detecting "normal" overflows, but not for scenarios like `MinValue..MaxValue..MaxValue`, [etc.](https://github.com/dotnet/fsharp/blob/fa4dcfcfd5afe71ba9f4f2d5269cb3c7b07536d8/tests/FSharp.Core.UnitTests/FSharp.Core/PrimTypes.fs#L831-L932), where the `step` itself is big enough to cause overflow every time.

I didn't even notice, though, because the `RangeTests` weren't failing—they were just never finishing 🙃. (Maybe they had passed during some earlier phase, and, since they never subsequently "failed," I never suspected that I might have introduced a regression.)

The proper check is indeed the one in the existing implementation, where you always verify that $next > current$ (when going up) or $next < current$ (when going down).

Reply:
So in e44dc37 I basically just decided to precompute the iteration count and handle any overflow scenarios there instead of needing to check for overflow on every iteration. That had the added benefit of letting me reuse the same logic for `for`-loops, list-initialization, and array-initialization (where we already needed to precompute the count anyway).

The basic logic looks something like this:

```fsharp
let count =
    if step = 0 then
        ignore ((.. ..) start step finish) // Throws.
    if 0 < step then
        if finish < start then 0 else unsigned (finish - start) / unsigned step + 1
    else // step < 0
        if start < finish then 0 else unsigned (start - finish) / (if step = MinValue then unsigned MaxValue + 1 else unsigned (abs step)) + 1
```

Where possible, some or all of this is done at compile-time so that we can emit less code.

Both the compile-time (inlined constants) and runtime (no inlined constants) versions of [these tests](https://github.com/dotnet/fsharp/blob/195ae177c226ad81e88f449a1b69593c7cf48a54/tests/FSharp.Core.UnitTests/FSharp.Core/PrimTypes.fs#L1005-L1083) now pass.

Reply:
This is really good, thanks for fixing this.
Without this, I think the range computation was an infinite loop.

Let's see what kind of Azure bill is sent for this :-D